File: Data Structures & Algorithms/meeting-schedule-ii/submission-3.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        vector<int> start;
        vector<int> end;
        for(auto i: intervals){
            start.push_back(i.start);
            end.push_back(i.end);
        }
        sort(start.begin(), start.end());
        sort(end.begin(), end.end());
        int s = 0;
        int e = 0;
        int res = 0;
        int curr = 0;
        while(s < start.size()){
            if(start[s] < end[e]){
                curr++;
                s++;
            }else{
                curr--;
                e++;
            }
            res = max(curr, res);
        }
        return res;

    }
};

File: Data Structures & Algorithms/insert-new-interval/submission-3.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        begin = newInterval[0]
        end = newInterval[1]
        for i in range(len(intervals)):
            curr = intervals[i]
            if curr[1] < begin:
                res.append(curr)
            elif curr[0] > end:
                res.append([begin, end])
                begin = curr[0]
                end = curr[1]
            else:
                begin = min(begin, curr[0])
                end = max(end, curr[1])
        res.append([begin, end])    
        return res